snap/quota/resources.go

```diff
@@ -135,38 +152,93 @@ func (qr *Resources) ValidateChange(newLimits Resources) error {
 		}
 	}
 
+	// Check that the cpu limit is not being removed, we do not support setting these
+	// two settings individually. Count/Percentage must be updated in unison.
+	if newLimits.CPU != nil && qr.CPU != nil {
+		if newLimits.CPU.Count == 0 && qr.CPU.Count != 0 {
+			return fmt.Errorf("cannot remove cpu limit from quota group")
+		}
+		if newLimits.CPU.Percentage == 0 && qr.CPU.Percentage != 0 {
+			return fmt.Errorf("cannot remove cpu limit from quota group")
+		}
+	}
+
+	// Check that we are not removing the entire cpu set
+	if newLimits.CPUSet != nil && qr.CPUSet != nil {
+		if len(newLimits.CPUSet.CPUs) == 0 {
+			return fmt.Errorf("cannot remove all allowed cpus from quota group")
+		}
+	}
+
+	// Check that the thread limit is not being decreased
+	if newLimits.Threads != nil {
+		if newLimits.Threads.Limit == 0 {
+			return fmt.Errorf("cannot remove thread limit from quota group")
+		}
+
+		// we disallow decreasing the thread limit initially until we understand
+		// the full consequences of doing so.
+		if qr.Threads != nil && newLimits.Threads.Limit < qr.Threads.Limit {
+			return fmt.Errorf("cannot decrease thread limit, remove and re-create it to decrease the limit")
+		}
+	}
+
 	return nil
 }
 
-// Change updates the current quota limits with the new limits. Additional verification
-// logic exists for this operation compared to when setting initial limits. Some changes
-// of limits are not allowed.
-func (qr *Resources) Change(newLimits Resources) error {
-	if err := qr.ValidateChange(newLimits); err != nil {
-		return err
+// clone returns a deep copy of the resources.
+func (qr *Resources) clone() Resources {
```

Comment:
I like this new approach, may a test like:
```go
diff --git a/snap/quota/export_test.go b/snap/quota/export_test.go
index 0d7ad423d1..47da6f2109 100644
--- a/snap/quota/export_test.go
+++ b/snap/quota/export_test.go
@@ -30,3 +30,7 @@ func (grp *Group) InspectInternalQuotaAllocations() map[string]*GroupQuotaAlloca
        grp.getQuotaAllocations(allQuotas, nil)
        return allQuotas
 }
+
+func ResourcesClone(r *Resources) Resources {
+       return r.clone()
+}
diff --git a/snap/quota/resources_test.go b/snap/quota/resources_test.go
index 45ece65bb6..4aac973182 100644
--- a/snap/quota/resources_test.go
+++ b/snap/quota/resources_test.go
@@ -20,6 +20,8 @@
 package quota_test
 
 import (
+       "reflect"
+
        . "gopkg.in/check.v1"
 
        "github.com/snapcore/snapd/gadget/quantity"
@@ -169,3 +171,22 @@ func (s *resourcesTestSuite) TestQuotaChangeValidationPasses(c *C) {
                c.Check(t.limits, DeepEquals, t.newLimits)
        }
 }
+
+func (s *resourcesTestSuite) TestResourceCloneComplete(c *C) {
+       r := &quota.Resources{}
+       rv := reflect.ValueOf(r).Elem()
+       for i := 0; i < rv.NumField(); i++ {
+               fv := rv.Field(i)
+               ft := fv.Type()
+               nv := reflect.New(ft.Elem())
+               fv.Set(nv)
+       }
+
+       // clone and ensure there are no un-initialized members after the clone
+       r2 := quota.ResourcesClone(r)
+       rv = reflect.ValueOf(r2)
+       for i := 0; i < rv.NumField(); i++ {
+               fv := rv.Field(i)
+               c.Check(fv.IsNil(), Equals, false)
+       }
+}
```
(or similar) to ensure that the clone is really complete and we don't forget to add things when we extend the `Resources` struct?

Reply:
I pushed this to https://github.com/snapcore/snapd/pull/11545